### DISCUSSION TITLE
UAC2: choose feedback method

### DISCUSSION THREAD

**Discussion by mastupristi** [hathach/tinyusb#2460] [General]
hello,
in my project I have TDM at 48kHz and I process it sample by sample, so I have an interrupt every 20.83us and it is within this interrupt that I would need to take the data coming from usb to send it on TDM.

Obviously the clock of this DM is asynchronous to the USB clock so I need the feedback endpoint. The problem is that I have absolutely no idea which method I should choose from those available:

```
  AUDIO_FEEDBACK_METHOD_FREQUENCY_FIXED,
  AUDIO_FEEDBACK_METHOD_FREQUENCY_FLOAT,
  AUDIO_FEEDBACK_METHOD_FREQUENCY_POWER_OF_2,
```
I tried, unsuccessfully, to understand something from the comments in the sources (here https://github.com/hathach/tinyusb/blob/master/src/class/audio/audio_device.h#L470).

So the question is, in my use case what method should I choose and what do I need to provide in terms of functions?

Keep in mind that I also have a hw timer available that could be fed by the MCLK @24.576MHz. I read something about this in the comments in the code (`and the cycle count of f_m at time of the SOF interrupt (e.g. by use of a hardware counter)` here https://github.com/hathach/tinyusb/blob/master/src/class/audio/audio_device.h#L472). I can imagine that it might be useful but I didn't understand how.

And then is there an API that allows me to retrieve data from the usb directly from an interrupt handler?

regards
Max

**Comment by lijunru-hub** [comment]
Will there be any references on how to use the feedback later on?

**Comment by HiFiPhile** [comment]
Back from holiday...

> in my project I have TDM at 48kHz and I process it sample by sample, so I have an interrupt every 20.83us and it is within this interrupt that I would need to take the data coming from usb to send it on TDM.

I highly recommend you to use buffer and DMA for audio streaming applications, interrupt at this high frequency can easily be delayed by other interrupts and cause data loss.

> So the question is, in my use case what method should I choose and what do I need to provide in terms of functions?

I suggest you to first try #2328 which take care of feedback calculation inside the class.

Even if you are measuring MCLK with a precision hardware timer there will still be accumulated error and drift, which are not accounted by other feedback method.
![image](https://github.com/hathach/tinyusb/assets/4375114/84282fa1-2cc4-4d4d-9629-094a5268c521)

**Reply by mastupristi** [reply]
> I highly recommend you to use buffer and DMA for audio streaming applications, interrupt at this high frequency can easily be delayed by other interrupts and cause data loss.

I am perfectly aware of the risks involved. I can tell you that I actually use DMA but I need the interrupt every sample for a low latency requirement. The interrupt is at the highest priority supported by my processor (a Cortex M7 running at 600MHz or more) and it's the only one that has that priority.
I have been applying this way of doing things for some years now and on many products, in even "critical" applications ( do you know the sphere?), and we have never had any problems

> I suggest you to first try https://github.com/hathach/tinyusb/pull/2328 which take care of feedback calculation inside the class.

I have studied and tried this branch of yours. I took inspiration from it to calculate the feedback but at the application level.

I renew the question: Are there any APIs that allow me to write and read data directly to/from tinyusb buffers/fifos from an interrupt handler?

best regards
Max
